### ISSUE TITLE
DDP training on TPU is not working

### ISSUE THREAD

**Issue by Liyang90** [Lightning-AI/lightning#15878]
### Bug description

PyTorch-Lightning DDP training on TPU might be broken for “automatic\_optimization” mode. 

[“automatic\_optimization” mode](https://pytorch-lightning.readthedocs.io/en/stable/common/optimization.html#automatic-optimization) is the default mode for a LightningModule to be trained by the Trainer. Users only need to define the training forward pass in the module and the Trainer would automatically do the backward pass and optimization step correctly.

On TPU (PT/XLA), DDP is usually achieved by calling `xm.optimizer_step()` for optimization step, which would add a gradient all\_reduce op before calling `optimizer.step()`. In PyTorch-Lightning, it is done in [TPUPrecisionPlugin](https://github.com/Lightning-AI/lightning/blob/33e1f9345d13c8dbb9047ecebc94745c576035e7/src/pytorch_lightning/plugins/precision/tpu.py#L43) class as:

```
closure_result = xm.optimizer_step(optimizer, optimizer_args={"closure": closure, **kwargs})
```

However, in “automatic\_optimization” mode, PyTorch-Lightning actually puts the forward pass and backward pass into the `closure` callable given to the `optimizer.step()`, so forward and backward happen within `optimizer.step()`. What ended up happening is, in an iteration of a batch:

1. All\_reduce gradients
2. Forward pass
3. Zero gradients
4. Backward pass
5. Optimizer step

The all\_reduce would be a no\_op, and the gradients are not being synchronized between DDP processes before being applied to the models. What should happen instead is:

1. Forward pass
2. Zero gradients
3. Backward pass
4. All\_reduce gradients
5. Optimizer step

The gradient all\_reduce op needs to be inserted into the closure. A possible fix could be:

```diff
diff --git a/src/pytorch_lightning/plugins/precision/tpu.py b/src/pytorch_lightning/plugins/precision/tpu.py
index efa61dd8f..2dda36d3f 100644
--- a/src/pytorch_lightning/plugins/precision/tpu.py
+++ b/src/pytorch_lightning/plugins/precision/tpu.py
@@ -29,6 +29,13 @@ class TPUPrecisionPlugin(PrecisionPlugin):
            raise ModuleNotFoundError(str(_XLA_AVAILABLE))
        super().__init__(*args, **kwargs)
+    def _tpu_wrap_closure(self, optimizer, closure: Callable[[], Any]) -> Any:
+        import torch_xla.core.xla_model as xm
+
+        closure_result = closure()
+        xm.reduce_gradients(optimizer)
+        return closure_result
+
    def optimizer_step(  # type: ignore[override]
        self,
        optimizer: Optimizable,
@@ -39,8 +46,9 @@ class TPUPrecisionPlugin(PrecisionPlugin):
    ) -> Any:
        import torch_xla.core.xla_model as xm
+        closure = partial(self._tpu_wrap_closure, optimizer, closure)
        closure = partial(self._wrap_closure, model, optimizer, optimizer_idx, closure)
-        closure_result = xm.optimizer_step(optimizer, optimizer_args={"closure": closure, **kwargs})
+        closure_result = optimizer.step(closure=closure, **kwargs)
        skipped_backward = closure_result is None
        # in manual optimization, the closure does not return a value
        if model.automatic_optimization and skipped_backward:
```

A comparison between with and without the patch is done on the modified [MNIST TPU tutorial](https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/mnist-tpu-training.html), with `BATCH_SIZE = 128`: [Tensorboard log](https://tensorboard.dev/experiment/Gj11qzUxRomNK224L7qdjA/). From the training loss curve in it, the model converges faster with the fixed codes, as the gradients are correctly reduced and models are synced across processes.

IR graphs can also be dumped before and after the fix. Before the fix, a `xla::cross_replica_sum()` op cannot be found in an iteration, while after the fix, it correctly appears after the backward ops.

### How to reproduce the bug

[MNIST TPU tutorial](https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/mnist-tpu-training.html)

### Error messages and logs

```

# Error messages and logs here please

```

### Environment

```

#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow): PyTorch-Lightning
#- PyTorch Lightning Version (e.g., 1.5.0): master
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10): 1.14
#- Python version (e.g., 3.9):
#- OS (e.g., Linux): Linux
#- CUDA/cuDNN version: None
#- TPU models and configuration: Google Cloud TPU V3-8
#- How you installed Lightning(`conda`, `pip`, source): source
#- Running environment of LightningApp (e.g. local, cloud): cloud

```

**Comment by carmocca** [comment]
Thank you so much! Do you want to open a PR implementing your solution?

**Comment by carmocca** [comment]
Looking at the source code (https://github.com/pytorch/xla/blob/6e6bb07e696c51a555a7d33433508ba236703a35/torch_xla/core/xla_model.py#L1027-L1032). It'd be better to avoid the duplicated reduction by not using `xm.optimizer_step` at all, and instead calling it's internal calls in our desired order

```diff
diff --git a/src/pytorch_lightning/plugins/precision/tpu.py b/src/pytorch_lightning/plugins/precision/tpu.py
index efa61dd8f..30c64751b 100644
--- a/src/pytorch_lightning/plugins/precision/tpu.py
+++ b/src/pytorch_lightning/plugins/precision/tpu.py
@@ -29,6 +29,13 @@ class TPUPrecisionPlugin(PrecisionPlugin):
             raise ModuleNotFoundError(str(_XLA_AVAILABLE))
         super().__init__(*args, **kwargs)
 
+    def _wrap_closure(self, model: "pl.LightningModule", optimizer: Optimizable, **kwargs: Any) -> Any:
+        closure_result = super()._wrap_closure(model, optimizer, **kwargs)
+        import torch_xla.core.xla_model as xm
+
+        xm.reduce_gradients(optimizer)
+        return closure_result
+
     def optimizer_step(  # type: ignore[override]
         self,
         optimizer: Optimizable,
@@ -37,10 +44,8 @@ class TPUPrecisionPlugin(PrecisionPlugin):
         closure: Callable[[], Any],
         **kwargs: Any,
     ) -> Any:
-        import torch_xla.core.xla_model as xm
-
         closure = partial(self._wrap_closure, model, optimizer, optimizer_idx, closure)
-        closure_result = xm.optimizer_step(optimizer, optimizer_args={"closure": closure, **kwargs})
+        closure_result = optimizer.step(closure=closure, **kwargs)
         skipped_backward = closure_result is None
         # in manual optimization, the closure does not return a value
         if model.automatic_optimization and skipped_backward:
```

**Comment by Liyang90** [comment]
Shouldn't grad clipping be done after grad all_reduce?

**Comment by carmocca** [comment]
Yes. My bad!

**Comment by Liyang90** [comment]
> Thank you so much! Do you want to open a PR implementing your solution?

I notice there are other places that have called `xm.optimizer_step()`, such as in lightning, lightning_lite, and pytorch_lightning. I don't know enough how to incorporate this change to all the places that have similar issue. So I hope someone know more about the package structure could help.

**Comment by carmocca** [comment]
AFAIK Lite does not suffer from this issue because it doesn't need to run the closure separately.

**Comment by Liyang90** [comment]
The fix is merged to master.
